File: src/wwo.cpp
#include "wwo.hpp"

double* WWO::n_array (int d, int x) {
  double *arr = new double[d];
  for (int i=0; i < d; i++) {
    arr[i] = x;
  }

  return arr;
}

double** WWO::n_matrix (int m, int n, double x) {
  double **mat = new double*[m];
  for (int i=0; i < m; i++) {
    mat[i] = new double[n];
    for (int j=0; j < n; j++) {
      mat[i][j] = x;
    }
  }

  return mat;
}

double WWO::nrand () {
  return ((double) std::rand() / __INT_MAX__);
}

double** WWO::init (int size, int dimension, double *lowers, double *uppers, double (*func)(double[], int)) {
  const int hMax = 12;
  double lambda = 0.5;

  double **x = n_matrix(size, dimension + 2, 0.0);

  for (int i=0; i < size; i++) {
    for (int j=0; j < dimension; j++) {
      x[i][j] = lowers[j] + nrand() * (uppers[j] - lowers[j]);
    }

    x[i][dimension] = func(x[i], dimension);
    x[i][dimension + 1] = hMax;
    x[i][dimension + 2] = lambda;
  }

  return x;
}

// Get the indicies of best and worst solutions of the population
int* WWO::minIndexMaxIndex (double **x, int size, int dimension) {
  int minIndex = 0;
  int maxIndex = 0;

  double bestValue = x[0][dimension];
  double worstValue = bestValue;

  for (int i = 1; i < size; i++) {
    if (x[i][dimension] < bestValue) {
      minIndex = i;
      bestValue = x[i][dimension];
    } else if (x[i][dimension] > worstValue) {
      maxIndex = i;
      worstValue = x[i][dimension];
    }
  }

  return new int[2]{minIndex, maxIndex};
}

int* WWO::randperm (int n, int k) {
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<> rand(1, k);

  int *result = new int[n];
  for (int i=0; i < n; i++) {
    result[i] = rand(gen);
  }

  return result;
}

double WWO::normrnd (double mu, double sigma) {
  std::default_random_engine generator;
  std::normal_distribution<double> distribution(mu, sigma);

  return distribution(generator);
}

void WWO::exec (int size, int dimension, double (*func)(double[], int), double lower, double upper) {
  // std::cout << "-> start" << std::endl;

  // lower bounds
  double *lowers = n_array(dimension, lower);

  // upper bounds
  double *uppers = n_array(dimension, upper);

  // number of function evaluations
  double nfes = dimension * 100000;

  double epsilon = 0.0000001;

  // parameter setting
  int hMax = 12;
  double alpha = 1.001;
  double betaMax = 0.25;
  double betaMin = 0.001;
  double beta = betaMax;
  int kmax = std::min(12, dimension / 2);

  // initialization
  // std::cout << "-> init" << std::endl;

  double **x = init(size, dimension, lowers, uppers, func);

  int *indexs = minIndexMaxIndex(x, size, dimension);
  int minIndex = indexs[0];
  int maxIndex = indexs[1];

  double optValue = x[minIndex][dimension];
  double optVector[dimension];
  std::copy(x[minIndex], x[minIndex] + dimension, optVector);

  std::cout << "-> optValue=" << optValue << std::endl;

  int nfe = 0;

  // iterative improvement
  // std::cout << "-> run" << std::endl;
  double *u = n_array(dimension, 0.0);

  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<double> unifrnd(-1, 1);

  while (nfe < nfes) {
    // std::cout << "-> nfe(" << nfe << ") " << std::endl;

    // Propagation
    for (int i=0; i < size; i++) {
      for (int d=0; d < dimension; d++) {
        u[d] = x[i][d] + unifrnd(gen) * x[i][dimension + 2] * (uppers[d] - lowers[d]);
        if (u[d] < lowers[d] || u[d] > uppers[d]) {
          u[d] = lowers[d] + nrand() * (uppers[d] - lowers[d]);
        }
      }

      double newValue = func(u, dimension);
      nfe += 1;

      if (newValue < x[i][dimension]) {
        std::copy(u, u + dimension, x[i]);
        x[i][dimension] = newValue;
        x[i][dimension + 1] = hMax;

        // Breaking
        if (newValue < optValue && i != minIndex) {
          optValue = newValue;
          int k = kmax;
          int *temp = randperm(dimension, k);

          for (int l=0; l < k; l++) {
            double tempX[dimension];
            std::copy(u, u + dimension, tempX);
            int d = temp[l];
            tempX[d] = x[i][d] * normrnd(0, 1) * beta * (uppers[d] - lowers[d]);
            if (tempX[d] < lowers[d] || tempX[d] > uppers[d]) {
              tempX[d] = lowers[d] + nrand() * (uppers[d] - lowers[d]);
            }

            double newValueTemp = func(tempX, dimension);
            nfe += 1;

            if (newValueTemp < newValue) {
              std::copy(tempX, tempX + dimension, x[i]);
              x[i][dimension + 2] *= newValueTemp / x[i][dimension];
              x[i][dimension] = newValueTemp;
              newValue = newValueTemp;

              if (newValueTemp < optValue) {
                optValue = newValueTemp;
                std::copy(tempX, tempX + dimension, optVector);
              }
            }
          }
        }
      } else {
        // decrease wave height
        x[i][dimension + 1] = x[i][dimension + 1] - 1;

        // Refraction
        if (x[i][dimension + 1] == 0) {
          for (int d=0; d < dimension; d++) {
            u[d] = normrnd((optVector[d] + x[i][d]) / 2, std::abs((optVector[d] - x[i][d]) / 2));
            if (u[d] < lowers[d] || u[d] > uppers[d]) {
              u[d] = lowers[d] + nrand() * (uppers[d] - lowers[d]);
            }
          }

          double oldValue = x[i][dimension];
          std::copy(u, u + dimension, x[i]);
          x[i][dimension] = func(u, dimension);
          nfe += 1;
          x[i][dimension + 1] = hMax;
          x[i][dimension + 2] *= x[i][dimension + 1] / oldValue;
        }
      }
    }

    // Update wavelenghts
    indexs = minIndexMaxIndex(x, size, dimension);
    minIndex = indexs[0];
    maxIndex = indexs[1];

    optValue = x[minIndex][dimension];
    std::copy(x[minIndex], x[minIndex] + dimension, optVector);

    double worstValue = x[maxIndex][dimension];
    double den = (worstValue - optValue + epsilon);
    for (int i=0; i < size; i++) {
      x[i][dimension + 2] *= std::pow(alpha, - (worstValue - x[i][dimension] + epsilon) / den);
    }

    beta = betaMax - (betaMax - betaMin) * nfe / nfes;
  }

  indexs = minIndexMaxIndex(x, size, dimension);
  minIndex = indexs[0];
  maxIndex = indexs[1];

  optValue = x[minIndex][dimension];
  std::copy(x[minIndex], x[minIndex] + dimension, optVector);

  std::cout << "-> optVector = [";
  for (int i=0; i < dimension; i++) {
    std::cout << optVector[i] << ((i == dimension-1) ? "" : ", ");
  }

  std::cout << "]" << std::endl;
  std::cout << "-> optValue = " << optValue << std::endl;
  std::cout << std::endl;
}

File: python/wwo.py
from cec2019comp100digit import cec2019comp100digit
from joblib import Parallel, delayed
import multiprocessing
import numpy as np
import random
import sys
import math
import time

def SphereFunc (x):
  return np.sum(np.power(x, 2))

def AckleyFunc (x, a=20, b=0.2, c=2*np.pi):
    d = len(x) # dimension of input vector x
    sum_sq_term = -a * np.exp(-b * np.sqrt(sum(x*x) / d))
    cos_term = -np.exp(sum(np.cos(c*x) / d))
    return a + np.exp(1) + sum_sq_term + cos_term

def evalFunc (x, id, dimension):
  bench = cec2019comp100digit
  bench.init(id, dimension)

  return bench.eval(x)
  bench.end()

def WWO (size, id, dimension, nfes = None, lowers = None, uppers = None):
  # constant
  if (lowers is None):
    lowers = np.full(dimension, -100.0)
  if (uppers is None):
    uppers = np.full(dimension, 100.0)
  if (nfes is None):
    nfes = dimension * 10000

  # parameter setting
  hMax = 12
  alpha = 1.001
  betaMax = 0.25
  betaMin = 0.001
  beta = betaMax
  kmax = min(12, dimension // 2)

  # initialiation
  x = Init(size, id, dimension, lowers, uppers)
  [minIndex, maxIndex] = MinIndexMaxIndex(x)
  optValue = x[minIndex][dimension]
  optVector = x[minIndex][:dimension]
  nfe = 0

  # print("-> optValue =", optValue)

  stTime = time.time()
  its = 0

  # iterative improvement
  u = np.zeros(dimension)
  while (nfe < nfes):
    its += 1
    for i in range(0, size):
      for d in range(0, dimension):
        u[d] = x[i][d] + (random.uniform(-1.0, 1.0) * x[i][dimension + 2] * (uppers[d] - lowers[d]))
        if (u[d] < lowers[d] or u[d] > uppers[d]):
          u[d] = lowers[d] + random.random() * (uppers[d] - lowers[d])
      newValue = evalFunc(u, id, dimension)
      nfe = nfe + 1
      if (newValue < x[i][dimension]):
        x[i][:dimension] = u.copy()
        x[i][dimension] = newValue
        x[i][dimension + 1] = hMax
        # Breaking
        if (newValue < optValue and i != minIndex):
          optValue = newValue
          k = kmax
          temp = np.random.permutation(dimension)[k:]
          for l in range(0, k):
            tempX = u.copy()
            d = temp[l]
            linearX = np.reshape(x, np.size(x), order='F').copy()
            tempX[d] = linearX[d] + random.normalvariate(0, 1) * beta * (uppers[d] - lowers[d])
            if (tempX[d] < lowers[d] or tempX[d] > uppers[d]):
              tempX[d] = lowers[d] + random.random() * (uppers[d] - lowers[d])
            newValueTemp = evalFunc(tempX, id, dimension)
            nfe = nfe + 1
            if (newValueTemp < newValue):
              linearX[d] = tempX[d]
              x = np.reshape(linearX, np.shape(x), order='F').copy()
              x[i][dimension + 2] = x[i][dimension + 2] * newValueTemp / x[i][dimension]
              x[i][dimension] = newValueTemp
              newValue = newValueTemp
              if (newValueTemp < optValue):
                optValue = newValueTemp
                optVector = x[i][:dimension]
      else:
        # decrease wave height
        x[i][dimension + 1] -= 1
        # Refraction
        if (x[i][dimension + 1] == 0):
          for d in range(0, dimension):
            u[d] = random.normalvariate((optVector[d]+x[i][d])/2, abs(optVector[d] - x[i][d])/2)
            if (u[d] < lowers[d] or u[d] > uppers[d]):
              u[d] = lowers[d] + random.random() * (uppers[d] - lowers[d])
          oldValue = x[i][dimension]
          x[i][:dimension] = u.copy()
          x[i][dimension] = evalFunc(u, id, dimension)
          nfe = nfe + 1
          x[i][dimension + 1] = hMax
          x[i][dimension + 2] = x[i][dimension + 2] * x[i][dimension] / oldValue

    [minIndex, maxIndex] = MinIndexMaxIndex(x)
    worstValue = x[maxIndex][dimension]
    optValue = x[minIndex][dimension]
    optVector = x[minIndex]
    den = (worstValue - optValue + sys.float_info.epsilon)
    for i in range(0, size):
      x[i][dimension + 2] = x[i][dimension + 2] * pow(alpha, -(worstValue-x[i][dimension]+sys.float_info.epsilon) / den)
    beta = betaMax - (betaMax - betaMin) * nfe / nfes

  [minIndex, maxIndex] = MinIndexMaxIndex(x)
  optValue = x[minIndex][dimension]
  optVector = x[minIndex]

  execTime = time.time()-stTime
  return [id, optValue, execTime, its]

def Init(size, id, dimension, lowers, uppers):
  hmax = 12
  wlambda = 0.5
  x = np.zeros((size, dimension + 3))

  for i in range(0, size):
    for j in range(0, dimension):
      x[i][j] = lowers[j] + random.random() * (uppers[j] - lowers[j])
    x[i][dimension] = evalFunc(x[i], id, dimension)
    x[i][dimension + 1] = hmax
    x[i][dimension + 2] = wlambda

  return x

# Get the indicies of best and worst solutions of the population
def MinIndexMaxIndex (x):
  [m, n] = np.shape(x)
  minIndex = 0
  maxIndex = 0
  bestValue = x[0][n - 3]
  worstValue = bestValue

  for i in range(1, m):
    if (x[i][n - 3] < bestValue):
      minIndex = i
      bestValue = x[i][n - 3]
    elif (x[i][n - 3] > worstValue):
      maxIndex = i
      worstValue = x[i][n - 3]

  return [minIndex, maxIndex]

num_cores = multiprocessing.cpu_count()
def RunValidation (id=1, runs=51):
  print("Running ", runs, " times function (", id, ")")
  results = results = Parallel(n_jobs=num_cores)(delayed(WWO)(5, id, 10, nfes=150000) for i in range(runs))
  results.append(WWO(5, id, 10, nfes=150000))
  # for i in range(1, runs+1):
  #   print("Running Function(", id, ")", i, "of", runs, "times...")
  #   results.append(WWO(5, id, 10, nfes=150000))

  print(results)
  return results

def SaveReport (file):
  results = np.loadtxt(open(file, "rb"), delimiter=",")
  stats = []
  for i in  range(len(results) // 51):
    d = 52 * i
    result = results[d:d+52]
    stat = [
      np.min(result[:,1]),
      np.max(result[:,1]),
      np.average(result[:,1]),
      np.std(result[:,1])
    ]
    stats.append(stat)

  print(stats)
  np.savetxt("report.csv", stats, delimiter=",")

print('WWO 5 population, 10 dimensions in ', num_cores, 'cores')
report = []
for i in range(1, 11):
  report.extend(RunValidation(i))
  np.savetxt("results_51.csv", report, delimiter=",")

SaveReport("results_51.csv")

# RunCECBenchmark()

# WWO(5, 1, 10)
# print('WWO 5 population, 30 dimensions')
# WWO(5, 1, 30)
